src/execution/operator/aggregate/physical_streaming_window.cpp

```diff
@@ -118,6 +101,98 @@ class StreamingWindowState : public OperatorState {
+
+		static bool ComputeOffset(ClientContext &context, BoundWindowExpression &wexpr, int64_t &offset) {
+			offset = 1;
+			if (wexpr.offset_expr) {
+				if (wexpr.offset_expr->HasParameter() || !wexpr.offset_expr->IsFoldable()) {
+					return false;
+				}
+				auto offset_value = ExpressionExecutor::EvaluateScalar(context, *wexpr.offset_expr);
+				if (offset_value.IsNull()) {
+					return false;
+				}
+				Value bigint_value;
+				if (!offset_value.DefaultTryCastAs(LogicalType::BIGINT, bigint_value, nullptr, false)) {
+					return false;
+				}
+				offset = bigint_value.GetValue<int64_t>();
+			}
+
+			//	We can only support negative LEAD values and positive LAG values
+			if (wexpr.type == ExpressionType::WINDOW_LEAD) {
+				offset = -offset;
+			}
+			return 0 <= offset && idx_t(offset) < MAX_BUFFER;
+		}
+
+		static bool ComputeDefault(ClientContext &context, BoundWindowExpression &wexpr, Value &result) {
+			if (!wexpr.default_expr) {
+				result = Value(wexpr.return_type);
+				return true;
+			}
+
+			if (wexpr.default_expr && (wexpr.default_expr->HasParameter() || !wexpr.default_expr->IsFoldable())) {
+				return false;
+			}
+			auto dflt_value = ExpressionExecutor::EvaluateScalar(context, *wexpr.default_expr);
+			return dflt_value.DefaultTryCastAs(wexpr.return_type, result, nullptr, false);
+		}
+
+		LeadLagState(ClientContext &context, BoundWindowExpression &wexpr)
+		    : wexpr(wexpr), curr(wexpr.return_type), prev(wexpr.return_type, MAX_BUFFER),
+		      temp(wexpr.return_type, MAX_BUFFER) {
+			ComputeOffset(context, wexpr, offset);
+			ComputeDefault(context, wexpr, dflt);
+			prev.Reference(dflt);
+			prev.Flatten(MAX_BUFFER);
+		}
+
+		void Execute(ExecutionContext &context, DataChunk &input, Vector &result) {
+			ExpressionExecutor executor(context.client, *wexpr.children[0]);
+			executor.ExecuteExpression(input, curr);
+
+			//	Copy prev[MAX_BUFFER-offset, MAX_BUFFER] => result[0, offset]
+			idx_t source_count = MAX_BUFFER;
+			idx_t source_offset = source_count - idx_t(offset);
+			idx_t target_offset = 0;
+			VectorOperations::Copy(prev, result, source_count, source_offset, target_offset);
+			//	Copy curr[0, count-offset] => result[offset, count]
+			target_offset = idx_t(offset);
+			source_count = input.size() - target_offset;
+			source_offset = 0;
+			VectorOperations::Copy(curr, result, source_count, source_offset, target_offset);
+			// 	Copy curr[0, count] => prev[prev.count - count, prev.count]
+			source_count = input.size();
+			source_offset = 0;
+			target_offset = MAX_BUFFER - source_count;
+			if (target_offset) {
+				//	Shift down incomplete buffers
+				VectorOperations::Copy(prev, temp, MAX_BUFFER, source_count, 0);
+				VectorOperations::Copy(temp, prev, target_offset, 0, 0);
+				VectorOperations::Copy(curr, prev, source_count, source_offset, target_offset);
+			} else {
+				//	Overwrite
+				VectorOperations::Copy(curr, prev, source_count, source_offset, target_offset);
```

Comment:
Profiling this it seems like almost all of the time is going into these `Copy` methods - in particular the nullmask copying is expensive. Since we only have nulls at the beginning, adding this line seems to speed things up significantly as the nullmask copying can then be skipped:

```cpp
FlatVector::Validity(prev).Reset();
```

We should also at some point look into speeding up the nullmask copying particularly when the offsets are aligned (since then we can copy entire `int64_t` blocks instead of doing bit operations per row).

Reply:
I also have an optimization for `VectorOperations::Copy` here - https://github.com/duckdb/duckdb/pull/12618 - that would improve the current implementation (although maybe not as much as resetting the validity entirely)

Reply:
Yes that made it 2x faster!

The problem with the `Reset` is that the default may not be `NULL` - it may be some other constant. I'll add a test for that.

Reply:
It's not a problem - we overwrite the entire buffer. Makes a small difference.